### PR TITLE
Send messages efficiently

### PR DESCRIPTION
**Pull request by mcobrien**
Adds a facade `send(from, to, content)` that sends emails using the most efficient ESTMP extensions as described in https://github.com/HubSpot/NioSmtpClient/issues/19, and adds `RSET` for you if necessary.

Depending on server support and the provided `MessageContent`, `send` might use:

* Pipelining
* Chunking
* 8bitmime

It will automatically dot-stuff messages for you if chunking isn't supported.

A new property on `MessageContent` can describe the message as 7 bit, 8 bit or unknown and the facade uses this to determine the right method to send the message. Messages marked as unknown are sent via chunking or 8bitmime if possible, falling back to an efficient test of the email to see if it contains 8 bit characters.

This PR does not include support for dynamically rewriting 8 bit messages to quoted-printable or base64. I'll add that in a future PR.

Fixes https://github.com/HubSpot/NioSmtpClient/issues/19

@axiak

### CODE REVIEW

File: src/main/java/com/hubspot/smtp/client/SmtpSession.java [hunk 1]
```diff
@@ -61,7 +67,7 @@
       SmtpCommand.QUIT,
       SmtpCommand.NOOP);
 
-  private static final Supplier<Executor> SHARED_DEFAULT_EXECUTOR = Suppliers.memoize(SmtpSession::getSharedExecutor);
+  private static final Supplier<Executor> SHARED_DEFAULT_EXECUTOR = () -> Suppliers.memoize(SmtpSession::getSharedExecutor).get();
```
Comment (axiak) [inline src/main/java/com/hubspot/smtp/client/SmtpSession.java]: why was this necessary?

File: src/main/java/com/hubspot/smtp/client/SmtpSession.java [hunk 2]
```diff
@@ -82,6 +88,7 @@ private static ExecutorService getSharedExecutor() {
   private final Executor executor;
   private final CompletableFuture<Void> closeFuture;
   private final AtomicInteger chunkedBytesSent = new AtomicInteger(0);
+  private final AtomicBoolean requiresRset = new AtomicBoolean(false);
```
Comment (axiak) [inline src/main/java/com/hubspot/smtp/client/SmtpSession.java]: if you don't need atomic primitives, I would just use `volatile boolean`. Saves a few bytes and limits what your intended uses are

File: src/main/java/com/hubspot/smtp/client/SmtpSession.java [hunk 3]
```diff
@@ -142,6 +149,89 @@ public boolean isEncrypted() {
...
+
+    if (content.getEncoding() == MessageContentEncoding.SEVEN_BIT) {
+      return sendAs7Bit(from, to, content);
+    }
+
+    if (ehloResponse.isSupported(Extension.EIGHT_BIT_MIME)) {
+      return sendAs8BitMime(from, to, content);
+    }
+
+    if (content.count8bitCharacters() == 0) {
```
Comment (axiak) [inline src/main/java/com/hubspot/smtp/client/SmtpSession.java]: do you need this api? would a better API be `boolean has8BitCharacters()`? That would allow short circuit evaluation.

File: src/main/java/com/hubspot/smtp/client/SmtpSession.java (hunk 1 above)
Comment (mcobrien) [inline src/main/java/com/hubspot/smtp/client/SmtpSession.java]: I can remove it. I used a `java.util.function.Supplier` while I was working on this, but `Suppliers.memoize` returns a Guava `Supplier`, so this was to convert between the two.

File: src/main/java/com/hubspot/smtp/messages/CrlfTerminatingChunkedStream.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    int length = chunk.readableBytes();
+    if (!isEndOfInput()) {
+      return chunk;
+    }
+
+    if (length >= 2 && chunk.getByte(length - 2) == CR && chunk.getByte(length - 1) == LF) {
+      return chunk;
+    }
+
+    return allocator.compositeBuffer(2).addComponents(true, chunk, allocator.buffer(2).writeBytes(TRAILING_BYTES));
```
Comment (axiak) [inline src/main/java/com/hubspot/smtp/messages/CrlfTerminatingChunkedStream.java]: in this mode, who is responsible for the final `\r\n.`? Is that part of dot stuffing?

File: src/main/java/com/hubspot/smtp/client/SmtpSession.java (hunk 3 above)
Comment (mcobrien) [inline src/main/java/com/hubspot/smtp/client/SmtpSession.java]: This is to support choosing between quoted-printable and base64.

Comment (axiak) [inline src/main/java/com/hubspot/smtp/client/SmtpSession.java]: ah :ok_hand:

File: src/main/java/com/hubspot/smtp/messages/CrlfTerminatingChunkedStream.java (hunk 4 above)
Comment (mcobrien) [inline src/main/java/com/hubspot/smtp/messages/CrlfTerminatingChunkedStream.java]: After sending the `MessageContent` we send `EMPTY_LAST_CONTENT` which is an instance of `LastSmtpContent`. Netty's `SmtpRequestEncoder` adds the final `.\r\n` [when it processes that](https://github.com/netty/netty/blob/4.1/codec-smtp/src/main/java/io/netty/handler/codec/smtp/SmtpRequestEncoder.java#L78-L81). 

We could alternatively send `.\r\n` ourselves, but SmtpRequestEncoder also sets a `contentExpected` flag when `DATA` is sent, and this is only unset when `LastSmtpContent` is written.
